### PR TITLE
Adds local blockchain and integration testing sections 

### LINKED ISSUE #179: Add a basic integration test example to the docs
Using local blockchain.

### PR DESCRIPTION
**Pull request by ymekuria**
Closes
#179 

Adds sections on setting up a local blockchain, deploying locally, and writing integration tests to the "How to test a zkApp page".

### CODE REVIEW

File: pages/en/zkapps/how-to-test-a-zkapp.mdx [hunk 1]
```diff
@@ -37,8 +37,97 @@ Because we are using Jest, it's helpful to break down all functionality of your
...
+let Local = Mina.LocalBlockchain();
+Mina.setActiveInstance(Local);
+```
+
+#### Deploying a contract locally
+
+The mock Mina local blockchain contains test accounts that can be used to deploy the contract and pay transaction fees.
+
+```ts
+// test accounts that pays all the fees, and puts additional funds into the zkapp.
```
Comment (jasongitmail) [inline pages/en/zkapps/how-to-test-a-zkapp.mdx:54]: Comment seems a bit unclear

File: pages/en/zkapps/how-to-test-a-zkapp.mdx [hunk 2]
```diff
@@ -37,8 +37,97 @@ Because we are using Jest, it's helpful to break down all functionality of your
...
+The `Mina.LocalBlockchain()` method specifies a mock Mina ledger of accounts that runs locally. It contains logic for updating the ledger that can be used to test your smart contract locally.
+
+```ts
+let Local = Mina.LocalBlockchain();
+Mina.setActiveInstance(Local);
+```
+
+#### Deploying a contract locally
+
+The mock Mina local blockchain contains test accounts that can be used to deploy the contract and pay transaction fees.
```
Comment (jasongitmail) [inline pages/en/zkapps/how-to-test-a-zkapp.mdx:51]: This might read more clearly if the steps were connected together somehow--e.g "First, "; "Next, "; "Then" (or similar), to help walk them through it as instructions.

File: pages/en/zkapps/how-to-test-a-zkapp.mdx [hunk 3]
```diff
@@ -37,8 +37,97 @@ Because we are using Jest, it's helpful to break down all functionality of your
...
+let txn = await Mina.transaction(feePayer, () => {
+  Party.fundNewAccount(feePayer);
+  zkAppInstance.deploy({ zkappKey: zkAppPrivateKey});
+});
+txn.send();
+```
+
+#### Writing integration tests
+
+While Jest can be used to write both unit and integration tests, It is reccomended to create linear scripts for smart contract integration tests. The scripting worflow is easier to reason about when creating large interconnected tests. It also makes debugging easier and provides good error tracing.
```
Comment (jasongitmail) [inline pages/en/zkapps/how-to-test-a-zkapp.mdx:79]: Not sure I agree on this. I believe it's a better convention to teach people how to use Jest, otherwise their code could break and they won't know until they decide to run the script again and hopefully notice it

**Review by jasongitmail — CHANGES_REQUESTED** [review]

File: pages/en/zkapps/how-to-test-a-zkapp.mdx [hunk 4]
```diff
@@ -37,8 +37,97 @@ Because we are using Jest, it's helpful to break down all functionality of your
 
 For examples of existing tests, we recommend creating a template example using the Mina zkApp CLI via `zk project <name>` and examining the test file there. [You will see a basic example of a few tests](https://github.com/o1-labs/zkapp-cli/blob/main/templates/project-ts/src/Add.test.ts) that deploy and update the state on a smart contract using Jest.
 
+#### Creating a local blockchain
+
+The `Mina.LocalBlockchain()` method specifies a mock Mina ledger of accounts that runs locally. It contains logic for updating the ledger that can be used to test your smart contract locally.
```
Comment (iregina) [inline pages/en/zkapps/how-to-test-a-zkapp.mdx:42]: I read your commits out of context and was confused at first about what this is about. Would it make more sense to intro this with the use case / purpose for why someone would need to create a local blockchain? Are there an other use cases other than testing zkApps? 

Ex. "One way of testing your smart contract is to run it locally on a mock blockchain using built-in methods in the Mina zkApp CLI."

File: pages/en/zkapps/how-to-test-a-zkapp.mdx [hunk 5]
```diff
@@ -37,8 +37,97 @@ Because we are using Jest, it's helpful to break down all functionality of your
...
+  zkAppInstance.deploy({ zkappKey: zkAppPrivateKey});
+});
+txn.send();
+```
+
+#### Writing integration tests
+
+While Jest can be used to write both unit and integration tests, It is reccomended to create linear scripts for smart contract integration tests. The scripting worflow is easier to reason about when creating large interconnected tests. It also makes debugging easier and provides good error tracing. 
+
+ A well written script tests the flow of your smart contracts expected behavior, and verifies correctness of state updates. After you test the expected behavior, it's benifical to verify the preconditions of your methods and test edge cases of your smart contract. Below is an example of a basic integration test script.
```
Comment (iregina) [inline pages/en/zkapps/how-to-test-a-zkapp.mdx:81]: Is preconditions hyphenated? No, it's not: https://en.wikipedia.org/wiki/Precondition. TIL.

File: pages/en/zkapps/how-to-test-a-zkapp.mdx (hunk 3 above)
Comment (ymekuria) [reply pages/en/zkapps/how-to-test-a-zkapp.mdx:79]: We decided to use script files for the vm integration tests after the discussion on this [PR](https://github.com/o1-labs/snarkyjs/pull/295). The scripts are added to run in CI. We can recommend as convention to add the integration tests to Jest if you prefer.

File: pages/en/zkapps/how-to-test-a-zkapp.mdx (hunk 2 above)
Comment (ymekuria) [reply pages/en/zkapps/how-to-test-a-zkapp.mdx:51]: Good idea.

File: pages/en/zkapps/how-to-test-a-zkapp.mdx (hunk 3 above)
Comment (jasongitmail) [reply pages/en/zkapps/how-to-test-a-zkapp.mdx:79]: For the vm integration tests, anything is fine as long because it's hooked into CI.

But for this, our recommended convention is important b/c affects the dev experience and opens the door for more bugs in the zkApp ecosystem if we're not hooking tests into the test runner we've chosen.

In response to Gregor's comment on the linked issue, if a failed test needs 1,000+ lines for error output (!) and Jest can't show it, then we should create a separate issue to look into that. If you agree with him, could you open an issue for that on the CLI please, with a mention of how to reproduce or we keep an eye out to add that if it ever occurs, please?

Comment (ymekuria) [reply pages/en/zkapps/how-to-test-a-zkapp.mdx:79]: I agree we have to be careful about what recommendations and conventions we set for the community. I'm fine to include Jest in our docs as convention. I'll rewrite this portion with Jest as the recommendation and then create a seperate issue to address error output.

Comment (jasongitmail) [reply pages/en/zkapps/how-to-test-a-zkapp.mdx:79]: Thanks!

**Review by jasongitmail — APPROVED** [review]

**Review by jasongitmail — APPROVED** [review]
LGTM!
